### PR TITLE
Ignore git clone prefix when checking clipboard for source URL

### LINKED ISSUE #6260: Detect 'git clone' prefix and ignore
## Current behaviour

I copy a branch from gerrit code review server (git clone ssh://...), want to clone a new repo and got and error ( it tries to execute git clone git clone ... )

## Expected behaviour
The prefix git clone is detected automatically and will be ignored

## Steps to reproduce

Copy and paste a link from gerrit

## Did this work in previous version of GitExtensions

Current (3.xx)

## Environment

- Git Extensions 3.1.0.5877
- Build bb1687f6bd0450d0c1a2bae10a37c65493fc1294
- Git 2.19.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3416.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by nyankoframe**
Fixes #6260 

## Proposed changes
 
-  Detects whether clipboard contents contain a potential URL using regex; if so, clipboard contents are provided as-is allowing user to modify contents prior to clone operation.

## Test methodology 

- Manual testing
- Added new tests to PathUtilTest.IsUrl and PathUtilTest.CanBeGitURL

## Test environment(s) 

- GIT 2.31.0.windows.1
- Windows 10.0.19042.868

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/FormClone.cs [hunk 1]
```diff
@@ -112,6 +112,15 @@ protected override void OnRuntimeLoad(EventArgs e)
                     {
                         string text = Clipboard.GetText(TextDataFormat.Text) ?? string.Empty;
 
+                        // Remove git clone prefix, if any.
+                        // This could be extended in future with a regex.
+                        string gitCloneText = "git clone";
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormClone.cs:117]:
```suggestion
                        const string gitCloneText = "git clone";
```

File: GitUI/CommandsDialogs/FormClone.cs [hunk 2]
```diff
@@ -112,6 +112,15 @@ protected override void OnRuntimeLoad(EventArgs e)
...
+                        int gitClonePrefix = text.IndexOf(gitCloneText, StringComparison.InvariantCultureIgnoreCase);
+                        if (gitClonePrefix != -1)
+                        {
+                            text = text.Substring(gitClonePrefix + gitCloneText.Length + 1);
+                        }
+
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormClone.cs:123]: Currently, the check will be for any "git clone" match (is the substing check correct?)
Untested suggestion - this should be a regexp

```suggestion
                        if (text.StartsWith(gitCloneText, StringComparison.InvariantCultureIgnoreCase))
                        {
                            text = text.Substring(gitClonePrefix.Length);
                        }

```

**Comment by nyankoframe** [comment]
I've been looking at using a regex and can see two approaches:

1. Use a regex to remove any leading git commands (such as git clone, git checkout)
2. Use a regex to capture anything that looks to be a URL

It would make more sense to modify PathUtil.IsUrl if doing the second approach, but I'm not sure what sort of effect it'll have on other features.

File: GitUI/CommandsDialogs/FormClone.cs [hunk 3]
```diff
@@ -112,6 +112,14 @@ protected override void OnRuntimeLoad(EventArgs e)
                     {
                         string text = Clipboard.GetText(TextDataFormat.Text) ?? string.Empty;
 
+                        // Remove git clone prefix, if any.
+                        // This could be extended in future with a regex.
+                        const string gitCloneText = "git clone";
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormClone.cs:117]:
```suggestion
                        const string gitCloneText = "git clone ";
```

File: GitUI/CommandsDialogs/FormClone.cs [hunk 4]
```diff
@@ -112,6 +112,14 @@ protected override void OnRuntimeLoad(EventArgs e)
...
+                        if (text.StartsWith(gitCloneText, StringComparison.InvariantCultureIgnoreCase))
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormClone.cs:118]: I don't believe git commands are case insensitive. Whilst you can call the actual git command as "GIT" on Windows, git commands are all in the lower case:
image

**Review by RussKie — COMMENTED** [review]
I'd like to see tests for this as well.

File: GitUI/CommandsDialogs/FormClone.cs (hunk 4 above)
Comment (nyankoframe) [reply GitUI/CommandsDialogs/FormClone.cs]: Thank you for the feedback; I decided to use a regex to attempt to detect the URL and then allow the user to modify it prior to the clone operation.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Please do a clean [rebase and squash](https://github.com/gitextensions/gitextensions/wiki/How-To%3A-Squash-and-Rebase-your-changes), and it's good to go

**Comment by RussKie** [comment]
Thank you!

**Comment by RussKie** [comment]
If you wish these changes to go out in 3.5 (which will be released very soon) then please make a PR against 3.5 branch.

**Comment by nyankoframe** [comment]
> If you wish these changes to go out in 3.5 (which will be released very soon) then please make a PR against 3.5 branch.

Done, see #9116

**Pull request merged**